rfcs/0002-tracing-cache.md

```diff
@@ -0,0 +1,95 @@
+
+#### On-Disk Layer
+The In-Memory layer is not expected to remain in cache for very long in an effort to reduce memory consumption. An on-disk layer helps to prevent more calls to the fetch layer than is absolutely necessary.
+
+This on-disk layer would be implemented as a binary file store with a cache eviction policy over postgres. It is expected that the most recent blocks would be indexed the most frequently, so the scoring mechanism for the cache would probably be some mix between LRU, block recency, and block finality.
+
+#### Fetch layer
+The existing code in the `traces` function would be renamed to `fetch_traces` and would be the bottom layer of the cache. At first, this would exist like it does today. In the future this may be augmented to either delegate to fetching from a parity API or to calculate traces on demand using a local etherium node, depending on which resources are currently the bottleneck.
+
+#### Futures re-use
+In order to prevent concurrent loads of data from one cache layer to another (Eg: multiple concurrent calls to the underlying eth.web3 API with the same parameters) the hierarchical cache would have a mechanism to delegate multiple tasks to the same future with a NonReady state.
+
+### Binary Format
+
+Trace data is extremely large. Compressing the data directly impacts how many traces for blocks we can keep at any layer of the cache.
+
+Ideally our binary format would have the following properties:
+* Ability to skip records which are filtered out without parsing
+* Zero copy to allow for a subset of a record's fields to be read when deciding whether to filter it. (Alternatively, easily separated hot/cold storage)
+* Little to no overhead for the schema, as there are no optional fields and backwards compatibility can be amortized into the collection
+* Processable as a linear stream with minimal cache misses, ideally not requiring the entire file to be loaded at once.
+* Move redundant data outside of the Trace (eg: Block Number)
+
+JSON is the format the data is delivered in today. JSON is neither compressed nor fast: https://eng.uber.com/trip-data-squeeze/. JSON does not have any of the properties of our ideal format. The entirety of a JSON file needs to be parsed whether or not all of it is actually used. JSON's self-description and flexibility come at a large cost for data which has a schema. As much of the data in a trace is binary data or data which does not have an established representation in JSON, additional overheads are incurred in encoding representations of traces in the primitives that JSON supports.
+
+Flatbuffers seems like a good initial candidate here which makes reasonable tradeoffs between skip/zero-copy and compression ratios. Significant gains over Flatbuffers are possible with a custom format, but this is not likely a good place to spend time in the short term. 
+
+This is subject to change if it is decided that filtering/indexing should happen within postgres.
```

Comment:
I believe @leoyvens had some suggestions on storing only parts of each block's traces in the cache. Perhaps he can elaborate.

Regarding the size of traces: I've pulled traces for a fairly large block (~260 transactions) and they are 201kb large. Multiply that by ~9,000,000 blocks and you end up with ~1.8TB. So storing _all_ traces would require a lot of disk space and I/O. 

Each trace by itself does not have a crazy overhead:
```json
{
      "action": {
        "callType": "call",
        "from": "0x0681d8db095565fe8a346fa0277bffde9c0edbbf",
        "gas": "0x2d4b0",
        "input": "0xa9059cbb0000000000000000000000003f706d0da87ceef53ebf701a47fd38d5e4f1ed6700000000000000000000000000000000000000000000000000000000abd7a95e",
        "to": "0xdac17f958d2ee523a2206206994597c13d831ec7",
        "value": "0x0"
      },
      "blockHash": "0x030bf7a9e55f0178aa28c700c9346a3af9d3a7c513ab23f6a46d5b0b3496f0db",
      "blockNumber": 9115243,
      "result": {
        "gasUsed": "0x8729",
        "output": "0x"
      },
      "subtraces": 0,
      "traceAddress": [],
      "transactionHash": "0x15d6d3cad7cd2faa3b68330bd47672e569458b62ff46fe07dcb9dedf8b807b9e",
      "transactionPosition": 48,
      "type": "call"
    },
```
@olibearo can probably tell which parts we need here, I suspect we need most of these fields. What makes traces large is mainly the `action > input` and `action > output` fields; especially the inputs can contain the entire bytecode of a contract. I think it's also worth looking at the different trace types to decide what a good format to store them would be. I've seen `call`, `crate` and `reward` traces so far.

Developing our own binary file format for these feels like overkill. I'd rather see if we can narrow down what data we want to store and how we can fit that into a Postgres schema.

Reply:
Having the block as the cache key is the simplest and also guarantees that we only need to get the traces from Ethereum once for each block, but then if any trace in a block is needed all traces in it will be stored, which will bloat disk size with traces that are potentially never used. To balance this we could use a more fine grained key for the cache, for example `(block, signature, contract)` since the call triggers in a subgraph will always specify a signature and usually specify an address.

Reply:
> Developing our own binary file format for these feels like overkill

> ... 201kb large. Multiply that by ~9,000,000 blocks and you end up with ~1.8TB

Those statements seem to conflict, unless our goal is to only store a very small percentage of blocks. (More on this in a later comment)

> What makes traces large is mainly the action > input and action > output fields;

I think what makes the trace large is not so much the fields, but the overhead required to store the schema with each record, and the usage of base64 encoding for lack of a bytes type and u128 type. Just eyeballing it, it looks like somewhere between 50-85% of the space is wasted to those.

I don't suggest developing our own binary format. I do suggested storing the data using Flatbuffers instead of Json, if using blob storage in Postgres.

> fit that into a Postgres schema.
This is compelling. I don't know yet what feature set Postgres affords but it's likely just as good or better than FlatBuffers, and may give us better options about where to filter. I'll do some poking around and see what's available here.

Reply:
> we could use a more fine grained key for the cache, for example `(block, signature, contract)` since the call triggers in a subgraph will always specify a signature and usually specify an address.

My worry for using `trace_filter` instead of `trace_block` and having a fine grained cache is that we'll never actually get a cache hit, except when rolling back due to blocks being reverted. Even then, the cache would seem dubious.

This worry is based on the following ideas about how a Graph Node works, which may be wrong:

- When a Graph Node indexes a subgraph, it plays blocks forward in time and doesn't tend to revisit the same block for the same subgraph.
- It is only necessary to retrieve the traces once when indexing a block.
- It is possible but uncommon for multiple subgraphs to be based on the same Ethereum contract.

One possible approach to resolving the issue of increasing cache hits while not storing a block would be for the `EthereumTraceCache` to know the full set of contracts and addresses that are required by all of the subgraphs that are currently being indexed. This might be a code smell depending on how the code is structured and whether there would be any code required to keep these in sync. Here when any contract from a block is requested the cache would prefetch all the contracts and addresses needed from that block in the same request and store them separately in Postgres.

Reply:
What you're missing is that subgraphs get re-synced with small changes all the time, in this situation a fine grained cache would still get plenty of hits.

Reply:
> What you're missing is that subgraphs get re-synced with small changes all the time, in this situation a fine grained cache would still get plenty of hits.

There it is. I knew there must be something I was missing and this is key. What causes a re-sync? Code changes for a subgraph?

Reply:
Yes, a subgraph is identified by the IPFS hash of its contents, if any source file changes, the subgraph must sync again when deployed. Developing a subgraph involves a lot of re-deploying and waiting to sync.

Reply:
> Developing a subgraph involves a lot of re-deploying and waiting to sync.

Would it make sense then to have caching disabled for a production Graph Node?

Reply:
Not at all, most of our users do their development iterations on our production infra.